File: lista5/zad1.py
import time

def index_to_configuration(index, n):
    """
    Map index to configuration
    """
    # C = (1,2,3,4,5)  ==>  idx = 1 + 2*6 + 3*6^2 + 4*6^3 + 5*6^4
    conf = [0 for i in range(n)]
    for i in reversed([j for j in range(n)]):
        val = index // (n+1)**i
        conf[i] = val
        index -= val * (n+1)**i
    
    return conf

def configuration_to_index(conf):
    """
    Map configuration to index
    """
    index = 0
    n = len(conf)
    for i in range(n):
        index += conf[i] * (n+1)**i
    return index

def get_configurations(n):
    """
    Return list with bool values for all configurations for given 'n'
    Set safe configurations to True
    """
    configurations = [False for i in range((n+1)**n)]
    for safe_idx in [configuration_to_index(c) for c in [[i for j in range(n)] for i in range(n+1)]]:
        configurations[safe_idx] = True
    return configurations

def get_possible_transitions(conf):
    """
    Return list of all possible transitions from configuration 'conf'
     - Configurations as lists
    """
    transitions = []
    
    if conf[0] == conf[-1]:
        new_conf = [*conf]
        new_conf[0] = (new_conf[0] + 1) % (len(conf) + 1)
        transitions.append(new_conf)
    
    for i in range(1, len(conf)):
        if conf[i] != conf[i-1]:
            new_conf = [*conf]
            new_conf[i] = new_conf[i-1]
            transitions.append(new_conf)
    
    return transitions

def dfs(actual_idx, configurations, order, n):
    """
    Calculate dfs nodes order, and verify algorithm - verified if all 
        configurations are True
    """
    if configurations[actual_idx]:
        return
    
    conf = index_to_configuration(actual_idx, n)
    transitions = get_possible_transitions(conf)

    for c in transitions:
        i = configuration_to_index(c)
        if configurations[i]:
            continue
        dfs(i, configurations, order, n)

    order.append(actual_idx)
    configurations[actual_idx] = True

def simulation(n):
    order = []
    configurations = get_configurations(n)
    
    # calculate dfs order
    for idx in range(len(configurations)):
        dfs(idx, configurations, order, n)
    
    # max path len for each configuration -> if safe configuration then path len is set to -1
    path_len = [0 if i == True else -1 for i in get_configurations(n)]
    
    # calculate longest paths for each node
    for idx in order:
        conf = index_to_configuration(idx, n)
        transitions = get_possible_transitions(conf)

        path_len[idx] = 1
        for c in transitions:
            i = configuration_to_index(c)
            if path_len[i] >= path_len[idx]:
                path_len[idx] = path_len[i] + 1

    print(f'n = {n} \n  Valid configurations: {sum(configurations)}\n  All configurations: {len(configurations)}\n  Longest path: {max(path_len)}')


if __name__ == '__main__':
    n = 8
    t0 = time.time()
    simulation(n)
    print(f'Time: {time.time() - t0} s')

File: lista5/zad1.go
package main

import (
	"fmt"
	"math"
)

type edge struct {
	v1 int
	v2 int
}

func indexToConf(index int, n int) []int {
	conf := make([]int, n)

	for i := n - 1; i >= 0; i-- {
		val := index / int(math.Pow(float64(n+1), float64(i)))
		conf[i] = val
		index -= val * int(math.Pow(float64(n+1), float64(i)))
	}

	return conf
}

func confToIndex(conf []int) int {
	index := 0
	n := len(conf)
	for i := 0; i < n; i++ {
		index += conf[i] * int(math.Pow(float64(n+1), float64(i)))
	}
	return index
}

func getConfigurations(n int) []bool {
	return make([]bool, int(math.Pow(float64(n+1), float64(n))))
}

func getSafeConfigurations(n int) []int {
	confs := make([]int, n+1)
	safeConf := make([]int, n)
	for i := 0; i <= n; i++ {
		for j := 0; j < len(safeConf); j++ {
			safeConf[j] = i
		}
		safeIdx := confToIndex(safeConf)
		confs[i] = safeIdx
	}

	return confs
}

func getTransitions(conf []int) [][]int {
	transitions := make([][]int, 0)

	if conf[0] == conf[len(conf)-1] {
		newConf := make([]int, len(conf))
		_ = copy(newConf, conf)
		newConf[0] = (newConf[0] + 1) % (len(conf) + 1)
		transitions = append(transitions, newConf)
	}

	for i := 1; i < len(conf); i++ {
		if conf[i] != conf[i-1] {
			newConf := make([]int, len(conf))
			_ = copy(newConf, conf)
			newConf[i] = newConf[i-1]
			transitions = append(transitions, newConf)
		}
	}

	return transitions
}

func traverse(actualConf []int, configurations *[]bool, edges *[]edge) {
	idx := confToIndex(actualConf)
	if (*configurations)[idx] == true {
		return
	}
	(*configurations)[idx] = true

	transitions := getTransitions(actualConf)
	for _, conf := range transitions {
		i := confToIndex(conf)
		*edges = append(*edges, edge{idx, i})
		if (*configurations)[i] == true {
			continue
		}
		traverse(conf, configurations, edges)
	}
}

func convertListOfEdges(listOfEdges *[]edge, numOfVertices int) [][]int {
	edges := make([][]int, numOfVertices)

	for _, e := range *listOfEdges {
		edges[e.v2] = append(edges[e.v2], e.v1)
	}

	return edges
}

func isSafeConf(conf int, safeConfs []int) bool {
	for _, sc := range safeConfs {
		if sc == conf {
			return true
		}
	}
	return false
}

func getLongestPath(edges *[][]int, safeConfs []int, currIdx int, isRoot bool) int {
	if isSafeConf(currIdx, safeConfs) && !isRoot {
		return 0
	}

	pathLen := 0
	for _, v1 := range (*edges)[currIdx] {
		l := getLongestPath(edges, safeConfs, v1, false)
		if l > pathLen {
			pathLen = l
		}
	}

	return 1 + pathLen
}

func simulation(n int) {
	configurations := getConfigurations(n)
	edges := make([]edge, 0)

	for i := 0; i < len(configurations); i++ {
		conf := indexToConf(i, n)
		traverse(conf, &configurations, &edges)
	}

	safeConfs := getSafeConfigurations(n)
	convertedEdges := convertListOfEdges(&edges, len(configurations))

	longestPath := 0
	for _, sc := range safeConfs {
		pathLen := getLongestPath(&convertedEdges, safeConfs, sc, true)
		if pathLen > longestPath {
			longestPath = pathLen
		}
	}

	fmt.Println("Longest path: ", longestPath)
	// fmt.Println(configurations)
}

func main() {
	simulation(6)
}
